sources/api_client.py

```diff
@@ -0,0 +1,217 @@
+            params (Optional[Dict[str, Any]]): The original query parameters used in the current API request.
+            json (Optional[Dict[str, Any]]): The original JSON body of the current API request.
+
+        Returns:
+            tuple: A tuple containing the updated URL, query parameters, and JSON body to be used
+                for the next API request. These values are used to progress through the paginated data.
+        """
+        ...
+
+    @abstractmethod
+    def extract_records(self, response: Response) -> Any:
+        """
+        Extract the records data from the response.
+
+        Args:
+            response (Response): The response object from the API.
+
+        Returns:
+            Any: The extracted records data.
+        """
+        ...
+
+
+class BaseNextUrlPaginator(BasePaginator):
+    def prepare_next_request_args(self, url, params, json):
+        return self._next_reference, params, json
+
+
+class HeaderLinkPaginator(BaseNextUrlPaginator):
+    """A paginator that uses the 'Link' header in HTTP responses
+    for pagination.
+
+    A good example of this is the GitHub API:
+        https://docs.github.com/en/rest/guides/traversing-with-pagination
+    """
+
+    def __init__(self, links_next_key: str = "next") -> None:
+        """
+        Args:
+            links_next_key (str, optional): The key (rel ) in the 'Link' header
+                that contains the next page URL. Defaults to 'next'.
+        """
+        super().__init__()
+        self.links_next_key = links_next_key
+
+    def update_state(self, response: Response) -> None:
+        self.next_reference = response.links.get(self.links_next_key, {}).get("url")
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json()
+
+
+class JSONResponsePaginator(BaseNextUrlPaginator):
+    """A paginator that uses a specific key in the JSON response to find
+    the next page URL.
+    """
+
+    def __init__(self, next_key: str = "next", records_key: str = "results"):
+        """
+        Args:
+            next_key (str, optional): The key in the JSON response that
+                contains the next page URL. Defaults to 'next'.
+            records_key (str, optional): The key in the JSON response that
+                contains the page's records. Defaults to 'results'.
+        """
+        super().__init__()
+        self.next_key = next_key
+        self.records_key = records_key
+
+    def update_state(self, response: Response):
+        self.next_reference = response.json().get(self.next_key)
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json().get(self.records_key, [])
+
+
+class BearerTokenAuth(AuthBase):
+    def __init__(self, token: str):
+        self.token = token
+
+    def __call__(self, request):
+        request.headers["Authorization"] = f"Bearer {self.token}"
+        return request
+
+
+def join_url(base_url: str, path: str) -> str:
```

Comment:
should we use [urljoin](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) instead?